gc/base/standard/ConcurrentGC.cpp

```diff
@@ -2207,7 +2196,7 @@ MM_ConcurrentGC::concurrentMark(MM_EnvironmentStandard *env, MM_MemorySubSpace *
 			if(_stats->switchExecutionMode(CONCURRENT_ROOT_TRACING1, CONCURRENT_ROOT_TRACING2)) {
 				/* Signal threads for async callback */
 				bool signaledThreadsForAsyncCallback = false;
-				_cli->concurrentGC_collectRoots(env, CONCURRENT_ROOT_TRACING1, &_scanClassesMode, signaledThreadsForAsyncCallback, taxPaid);
```

Comment:
This state machine seems very language specific. 

The concurrent collector would know about roots and could inform the language it is time to scan roots and then it would be up to the language to implement a state machine to do different roots if it needs to do them at different time.